### DISCUSSION TITLE
The behavior of "vs" and "sp" is wrong is tags 1.31.0-beta.2

### DISCUSSION THREAD

**Discussion by xianyuIC** [NeoVintageous/NeoVintageous#917] [Q&A] [answered]
In version 1.30.1, with the addition of the "Origami" plugin, the commands for `:vs`, `:sp` and `:close` are all correct. Their behavior is consistent with that of vim, except that it does not recognize environment variables. For example,`:vs $HOME/test.v` cannot open an existing file.

But overall, the `:vs`,  `:sp` operations in version 1.30.1 are correct.

I tried version 'tags 1.31.0 beta. 2', and `:vs` and `:sp` can recognize environment variables, such as correctly opening `:vs $HOME/test. v`, but:

1. It cannot open files with absolute paths, such as `:vs ../ test2.v`.
(1) The `:vs` and `:sp` in version 1.30.1 can do it.
(2) The `gf` in version 1.31.0 beta.2 also cannot recognize absolute paths.

2. The window arrangement is incorrect. The behavior of `:vs` and `:sp` does indeed result in windows segmentation, but the vs opened file is in the old left window bar, and the right window bar is blank, it's wrong.
In version 1.30.1, their window arrangement is the correct behavior.

**Comment by gerardroche** [comment]
I can't reproduce these issues.

## 1. `:vs ../test2.v`

Is there a space in the path?

What is the current working directory? `:pwd`? 

For example, if the `:pwd` is `/home/code/project` and `test2.v` exists at ``/home/code/test2.v` then `:vs ../test2.v` and `gf` on the same path should work.

I'm guessing it's probably an issue with a space in the path?

What OS are you using?


## 2. window arrangement is incorrect

I don't understand the issue. Can you describe steps to reproduce?

**Comment by gerardroche** [comment]
fyi, the command line has auto complete. Press `<Tab>` to auto complete so... `:vs ../<Tab>` will show you what files are available, keep pressing `<Tab>` to cycle through the possible completions.

**Comment by xianyuIC** [comment]
## 1. `:vs ../test2.v`

my tree is like this :

```
test
|
|---- code
|          |
|          ---- test4.v
|---- test1.v
|---- test2.v
|---- test3.v
```

I entered the directory code and executed `sublime test4.v`, and I saw the following:

```
/home/ICer/test/test1.v
${HOME}/test/test2.v
../test3.v
```

### （1）gf

My cursor stopped at different positions to test `gf`. The first line succeeded, the second line succeeded, and the third line failed, error info is :  `E447: Cannot find file '../test3.v' in path. 

But when I open the test4. v file with vim and perform the same operation, all operations are correct.



### （2）vs

```
//it is OK
:vs /home/ICer/test/test1.v

//it is OK
:vs ${HOME}/test/test2.v

//it is wrong, It opened a new file, but it is not my existing test3.v file
:vs ../test3.v
```

I executed `pwd` in the command of sublime and was surprised to find that the displayed content was: `/home/subtitle/opt/subtitle_Text`. This is too strange.Sublime seems to have made an incorrect judgment about my execution directory...



In version 1.30.0, I executed `pwd` in the command of sublime and the displayed content was also : `/home/subtitle/opt/subtitle_Text`, but when i execute`:vs ../test3.v`，it is OK, it can open the correct test3.v file.



my OS is CentOS7.





## 2. window arrangement is incorrect

in 1.31.0-beta2，I entered the directory code and executed `sublime test4.v`,  then executed`:vs /home/ICer/test/test1.v` ,window is like this:

```
| test4.v | test1.v |            ||                             |
-----------------------------------------------------------------
|                                ||                             |
|                                ||                             |
|  this is test1                 ||                             |
|  this is test1                 ||                             |
|                                ||                             |
```

it is wrong window arrangement.



in 1.30.0 and vim, executed`:vs /home/ICer/test/test1.v` ,window is like this:

```
| test4.v |                      || test1.v |                   |
-----------------------------------------------------------------
|                                ||                             |
|                                ||                             |
|                                ||  this is test1              |
|                                ||  this is test1              |
|                                ||                             |
```

it is correct window arrangement.

**Comment by gerardroche** [comment]
## gf

The reason gf is not working is because the `:pwd` is pointing to the cwd set by Sublime Text on startup.  

NeoVintageous finds the cwd by checking:

1. the cwd set by the user e.g. `:cd ~/project`
2. if the cwd is not set, then it uses the root folder of an open project/folder
3. if there is no open project it defaults to the cwd set by sublime, which is the st installation folder

What we could do, is insert another check between 2 and 3:

1. the cwd set by the user e.g. `:cd ~/project`
2. if the cwd is not set, then it uses the root folder of an open project/folder
3. **if there is no open project, use the directory of the currently open file** 
4. if there is no opened file, then default to the cwd set by sublime, which is the st installation folder

I think this will be more confusing to users.  I think sublime should set the cwd when opening Sublime from the command line via `subl {file}`, but it does do that.  I think that's a bug/feature in ST.  I'm not sure there's anything I can do about that.

## vs

I can't reproduce this issue. Origami is required for this. It wasn't required in an earlier version, but to add the new features I had to use origami for all the window splitting and new window split ex commands.

If the issue still persists. Enable logging (see https://github.com/NeoVintageous/NeoVintageous/blob/master/.github/CONTRIBUTING.md) and post the console log.

You should see something like this:

```
NeoVintageous.DEBUG   [     ex_cmds.py:1310] do ex cmdline >>>:vs /home/code/path/to/file<<<
NeoVintageous.DEBUG   [      parser.py:79  ] parsing source >>>vs /home/code/path/to/file<<<
NeoVintageous.DEBUG   [      parser.py:89  ] parsed vsplit {'content': 'vsplit', 'name': 'vsplit', 'target': 'vsplit', 'params': {'file': '/home/code/path/to/file'}, 'forced': False, 'addressable': False, 'cooperates_with_global': False}
NeoVintageous.DEBUG   [     ex_cmds.py:1361] execute ex command: vsplit {'file': '/home/code/path/to/file', 'view': View(15)}
command: create_pane {"direction": "right"}
```

**Comment by xianyuIC** [comment]
## gf

I understand. Since there's no solution, let's forget it. In practical work, reports often contain `${HOME}` or other absolute paths, and code like `../test2. v` rarely appears, so it doesn't matter much if this cannot be done. Now 1.31.0-beta2 's  `gf` can recognize `${HOME}`, it's already great



## vs

#### (1) 1.30.0 version

`:vs ../test3.v` , it is OK，it opened my existing test3.v, the consolar content is ：

```
Vintageous: DEBUG   [     ex_cmds.py:1369] do ex cmdline >>>:vs ../test3.v<<<
NeoVintageous: DEBUG   [      parser.py:  79] parsing source >>>vs ../test3.v<<<
NeoVintageous: DEBUG   [      parser.py:  89] parsed vsplit {'cooperates_with_global': False, 'content': 'vsplit', 'target': 'vsplit', 'forced': False, 'params': {'file': '../test3.v'}, 'name': 'vsplit', 'addressable': False}
NeoVintageous: DEBUG   [     ex_cmds.py:1420] execute ex command: vsplit {'view': <sublime.View object at 0x7fbb0c69e0d0>, 'file': '../test3.v'}
command: show_panel {"panel": "console", "toggle": true}
```

`:sp ../test2.v` , it is OK，it opened my existing test2.v the consolar content is ：

```
NeoVintageous: DEBUG   [     ex_cmds.py:1369] do ex cmdline >>>:sp ../test2.v<<<
NeoVintageous: DEBUG   [      parser.py:  79] parsing source >>>sp ../test2.v<<<
NeoVintageous: DEBUG   [      parser.py:  89] parsed split {'cooperates_with_global': False, 'content': 'split', 'target': 'split', 'forced': False, 'params': {'file': '../test2.v'}, 'name': 'split', 'addressable': False}
NeoVintageous: DEBUG   [     ex_cmds.py:1420] execute ex command: split {'view': <sublime.View object at 0x7fbb0c6818d0>, 'file': '../test2.v'}
command: show_panel {"panel": "console", "toggle": true}
```

#### (2) 1.31.0-beta2 version

`:vs ../test3.v` , it is wrong，it opened a new test3.v in the path：`/home/sublime/opt/test3.v`, the consolar content is ：

```
NeoVintageous.DEBUG   [     ex_cmds.py:1310] do ex cmdline >>>:vs ../test3.v<<<
NeoVintageous.DEBUG   [      parser.py:79  ] parsing source >>>vs ../test3.v<<<
NeoVintageous.DEBUG   [      parser.py:89  ] parsed vsplit {'name': 'vsplit', 'params': {'file': '../test3.v'}, 'cooperates_with_global': False, 'target': 'vsplit', 'content': 'vsplit', 'forced': False, 'addressable': False}
NeoVintageous.DEBUG   [     ex_cmds.py:1361] execute ex command: vsplit {'view': <sublime.View object at 0x7f5906a3ed90>, 'file': '../test3.v'}
Unable to open /home/sublime/opt/test3.v
command: show_panel {"panel": "console", "toggle": true}
```

`:sp ../test3.v` , it is wrong，it opened a new test2.v in the path:`/home/sublime/opt/test2.v`, the consolar content is ：

```
NeoVintageous.DEBUG   [     ex_cmds.py:1310] do ex cmdline >>>:sp ../test2.v<<<
NeoVintageous.DEBUG   [      parser.py:79  ] parsing source >>>sp ../test2.v<<<
NeoVintageous.DEBUG   [      parser.py:89  ] parsed split {'name': 'split', 'params': {'file': '../test2.v'}, 'cooperates_with_global': False, 'target': 'split', 'content': 'split', 'forced': False, 'addressable': False}
NeoVintageous.DEBUG   [     ex_cmds.py:1361] execute ex command: split {'view': <sublime.View object at 0x7f5906a0af10>, 'file': '../test2.v'}
Unable to open /home/sublime/opt/test2.v
command: show_panel {"panel": "console", "toggle": true}
command: drag_select {"event": {"button": 1, "x": 647.9790
```



## window arrangement is incorrect

As for this issue, when we use `:vs` of `:sp`, can we fix the window arrangement back to 1.30.0? The window arrangement of 1.30.0 is correct.



By the way:
When testing version 1.30.0, I installed NeoVintageous directly using the sublime `install package` command, and later it was in the`~/.config/subtitle text-3/Installed Packages/NeoVintageous.sublime-package`.
When testing version 1.31.0 beta 2, I first uninstalled the version 1.30.0 using the sublime command `remove package`. Then download 1.31.0 beta 2 from github and place the entire NeoVintagous file in the `~/.config/sub lime text 3/Packages/`.

**Comment by gerardroche** [comment]
## gf

Remember you can manually set your cwd via `:cd {dir}`. And if you have a project or folder open neo will auto set the cwd to the root of you project.

## vs

This is the same issue as the gf issue. Notice the cwd path is `/home/sublime/opt`.

```
Unable to open /home/sublime/opt/test3.v
```

The reason `:vs` isn't working in your example (`:sp` never worked like this) is that `:vs` used the base directory of the current file as the cwd. For example, like I explained above, we could do:

> What we could do, is insert another check between 2 and 3:
>
>  1.  the cwd set by the user e.g. :cd ~/project
>  2.  if the cwd is not set, then it uses the root folder of an open project/folder
>  3. ** if there is no open project, use the directory of the currently open file**
>  4.  if there is no opened file, then default to the cwd set by sublime, which is the st installation folder
 
For example, if you try do the same with **no** files open it still won't work. It only worked because if you have `/path/to/file.txt` open and run `:vs ../file2.txt` it will use the cwd of the current file `/path/to/file.txt` so it will look for ``/path/to/../file2.txt`.

I think will add this behaviour to the ex open file commands because it's better than defaulting to Sublime's cwd, which is Sublime's installation directory and is useless.

## window arrangement 

I just realised the issue. You mean, 

- if there is an **empty** group then **don't** create another group, just reuse the empty one.
- if there is a non-empty, then create a new group

Is that correct? I think this makes sense.

Or do you mean
 
- if there is group then **don't** create another group, just reuse the existing group.

In Vim, a new window/group is always created so I don't like this.



> By the way:
> When testing version 1.30.0, I installed NeoVintageous directly using the sublime install package command, and later it was in the~/.config/subtitle text-3/Installed Packages/NeoVintageous.sublime-package.
> When testing version 1.31.0 beta 2, I first uninstalled the version 1.30.0 using the sublime command remove package. Then download > 1.31.0 beta 2 from github and place the entire NeoVintagous file in the ~/.config/sub lime text 3/Packages/.

When you install NeoVintageous it is unpacked from the `.sublime-package` into the `Packages` directory. This is normal.  When you install via a download or git clone you put in the `Packages` directory. This is normal too.

**Reply by xianyuIC** [reply]
## vs

No, you didn't understand what I meant.

my tree is like this :

```
test
|
|---- code
|          |
|          ---- test4.v
|---- test1.v
|---- test2.v
|---- test3.v
```

#### (1) 1.30.0 version

I entered the directory `test/code`, now I'm in the directory  `test/code` , then I executed `sublime test4.v`,

`:vs ../test3.v` , it is OK，it opened my existing test3.v, 

#### (2) 1.31.0-beta2 version

I entered the directory `test/code`, now I'm in the directory  `test/code` , then I executed `sublime test4.v`,

`:vs ../test3.v` , it is wrong，it opened a new test3.v in the path：`/home/sublime/opt/test3.v`

but the consolar content is : 

```
NeoVintageous.DEBUG   [     ex_cmds.py:1310] do ex cmdline >>>:vs ../test3.v<<<
NeoVintageous.DEBUG   [      parser.py:79  ] parsing source >>>vs ../test3.v<<<
NeoVintageous.DEBUG   [      parser.py:89  ] parsed vsplit {'name': 'vsplit', 'params': {'file': '../test3.v'}, 'cooperates_with_global': False, 'target': 'vsplit', 'content': 'vsplit', 'forced': False, 'addressable': False}
NeoVintageous.DEBUG   [     ex_cmds.py:1361] execute ex command: vsplit {'view': <sublime.View object at 0x7f5906a3ed90>, 'file': '../test3.v'}
Unable to open /home/sublime/opt/test3.v
command: show_panel {"panel": "console", "toggle": true}
```

I don't know why it displays `Unable to open /home/sublime/opt/test3.v`, acctually, it opened a new /home/sublime/opt/test3.v files.



**notice:** In both versions of the testing, I entered the directory `test/code`.  then I executed `sublime test4.v`,

* Version 1.30.0 was able to correctly open existing test3.v
* while version 1.31.0-beta2 opened `/home/subtitle/opt/test3.v`,  And the console printed:`Unable to open /home/sublime/opt/test3.v`. it open a wrong file, and the consolar message is also wrong. This is too strange.



## 2. window arrangement is incorrect

No, you didn't understand what I meant.

#### (1) 1.30.0 version

I entered the directory `test`, now I'm in the directory  `test` , then I executed `sublime test1.v`, the window is like this

```
| test1.v |                                                     |
-----------------------------------------------------------------
|                                                               |
|                                                               |
| this is test1                                                 |
| this is test1                                                 |
|                                                               |
```

I excute `:vs /home/ICer/test/test2.v`, the window is like this:

```
| test1.v |                      || test2.v |                   |
-----------------------------------------------------------------
|                                ||                             |
|                                ||                             |
| this is test1                  || this is test2               |
| this is test1                  || this is test2               |
|                                ||                             |
```

it is correct window arrangement.

#### (1) 1.31.0-beta2

I entered the directory `test`, now I'm in the directory  `test` , then I executed `sublime test1.v`, the window is like this

```
| test1.v |                                                     |
-----------------------------------------------------------------
|                                                               |
|                                                               |
| this is test1                                                 |
| this is test1                                                 |
|                                                               |
```

I excute `:vs /home/ICer/test/test2.v`, the window is like this:

```
| test1.v | test2.v |            ||                             |
-----------------------------------------------------------------
|                                ||                             |
|                                ||                             |
|  this is test2                 ||                             |
|  this is test2                 ||                             |
|                                ||                             |
```

it is wrong window arrangement.

The window is divided into two columns, but the right column is empty and has no files. The file test2.v opened as `vs` is in the left column and placed next to the label of test1. v.

it is wrong window arrangement.



The window arrangement in version 1.30.0 is correct.

**Comment by gerardroche** [comment, marked as answer]
Try that. beta.3

**Reply by xianyuIC** [reply]
## gf

I entered the directory `test/code`, now I'm in the directory  `test/code` , then I executed `sublime test4.v`,
---

gf in `../test1.v`    --> OK

gf  in `${HOME}/test/test1.v`  --> OK

gf in `/home/ICer/test/test1.v` --> OK

---

`:vs ../test1.v`    --> OK

`:vs ${HOME}/test/test1.v`  --> OK

`:vs /home/ICer/test/test1.v` --> OK

---

`:sp ../test1.v`    --> OK

`:sp ${HOME}/test/test1.v`  --> OK

`:sp /home/ICer/test/test1.v` --> OK

---

`:e ../test1.v`    --> OK

`:e ${HOME}/test/test1.v`  --> OK

`:e /home/ICer/test/test1.v` --> OK

---

`:pwd`, displays `/home/ICer/test/code`, It is correct !

---

All operations are correct! This is great !

The window arrangement is also correct !

Excellent !


Thank you very much,  You are very professional, thank you NeoVintagous !

**Reply by gerardroche** [reply]
Great. Thank you for your persistence. We got there in the end. :rocket:
